File: android/lib/src/main/kotlin/me/anasmusa/learncast/lib/screen/auth/TelegramLoginScreen.kt
package me.anasmusa.learncast.lib.screen.auth

import android.graphics.Bitmap
import android.os.Handler
import android.os.Looper
import android.webkit.JavascriptInterface
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.core.net.toUri
import me.anasmusa.learncast.core.appConfig

@Composable
fun TelegramLoginScreen(
    onGetResult: (result: String) -> Unit,
    onCancel: () -> Unit,
) {
    var isLoading by remember { mutableStateOf(true) }
    val urlToLoad = "https://oauth.telegram.org/auth?bot_id=${appConfig.telegramBotId}&origin=${appConfig.publicBaseUrl}&lang=uz"
    Box(
        modifier =
            Modifier
                .fillMaxWidth()
                .heightIn(min = 600.dp)
                .background(Color.White),
    ) {
        AndroidView(
            factory = { context ->
                WebView(context).apply {
                    setBackgroundColor(android.graphics.Color.WHITE)
                    settings.javaScriptEnabled = true
                    settings.domStorageEnabled = true
//                    val cookieManager = CookieManager.getInstance()
//                    cookieManager.removeAllCookies(null)
//                    cookieManager.flush()
                    addJavascriptInterface(
                        CancelJsInterface(onCancel),
                        "AndroidCancelHandler"
                    )

                    webViewClient =
                        object : WebViewClient() {
                            override fun onPageStarted(
                                view: WebView?,
                                url: String?,
                                favicon: Bitmap?,
                            ) {
                                super.onPageStarted(view, url, favicon)
                                isLoading = true
                            }

                            override fun onPageFinished(
                                view: WebView?,
                                url: String?,
                            ) {
                                super.onPageFinished(view, url)
                                isLoading = false
                                injectCancelInterceptor(this@apply)
                            }

                            override fun shouldOverrideUrlLoading(
                                view: WebView?,
                                request: WebResourceRequest?,
                            ): Boolean {
                                val url = request?.url.toString()

                                val uri = url.toUri()
                                val hash = uri.fragment // gives "tgAuthResult=eyJpZCI6..."
                                val jsonPart = hash?.removePrefix("tgAuthResult=")

                                if (jsonPart != null) onGetResult(jsonPart)
                                return false
                            }
                        }

                    loadUrl(urlToLoad)
                }
            },
            update = {
                it.loadUrl(urlToLoad)
            },
        )

        if (isLoading) {
            CircularProgressIndicator(
                modifier =
                    Modifier
                        .padding(top = 64.dp)
                        .align(Alignment.TopCenter),
                color = MaterialTheme.colorScheme.primary,
                strokeWidth = 3.dp,
            )
        }
    }
}

class CancelJsInterface(
    private val onCancel: () -> Unit
) {

    @JavascriptInterface
    fun onCancel() {
        Handler(Looper.getMainLooper()).post {
            this.onCancel.invoke()
        }
    }
}


private fun injectCancelInterceptor(webView: WebView) {
    val js = """
        (function() {
            const originalLoginCancel = window.loginCancel;

            window.loginCancel = function(event) {
                if (event) {
                    event.preventDefault();
                    event.stopPropagation();
                }

                AndroidCancelHandler.onCancel();

                if (originalLoginCancel && typeof originalLoginCancel === 'function') {
                    originalLoginCancel.call(this, event);
                }

                return false;
            };

            document.addEventListener('click', function(e) {
                const button = e.target.closest('button');

                if (
                    button &&
                    button.classList.contains('button-item-flat') &&
                    button.getAttribute('onclick') &&
                    button.getAttribute('onclick').includes('loginCancel')
                ) {
                    e.preventDefault();
                    e.stopPropagation();
                    AndroidCancelHandler.onCancel();
                }
            }, true);
        })();
    """.trimIndent()

    webView.evaluateJavascript(js, null)
}

File: ios/lib/screen/auth/TelegramLoginScreen.swift
//
//  TelegramLoginScreen.swift
//  ios
//
//  Created by Anas Erkinjonov on 31/01/26.
//

internal import Shared
import SwiftUI
import WebKit

struct TelegramLoginScreen: View {
    let onGetResult: (String) -> Void
    let onCancel: () -> Void

    @State private var isLoading = true

    private var urlToLoad: String {
        "https://oauth.telegram.org/auth?bot_id=\(appConfig.telegramBotId)&origin=\(appConfig.publicBaseUrl)&lang=uz"
    }

    var body: some View {
        ZStack(alignment: .top) {
            WebView(
                urlString: urlToLoad,
                isLoading: $isLoading,
                onGetResult: onGetResult,
                onCancel: onCancel
            )
            .frame(minHeight: 600)
            .background(Color.white)

            if isLoading {
                ProgressView()
                    .progressViewStyle(CircularProgressViewStyle(tint: .accentColor))
                    .scaleEffect(1.2)
                    .padding(.top, 64)
            }
        }
    }
}

// MARK: - WebView Wrapper
struct WebView: UIViewRepresentable {
    let urlString: String
    @Binding var isLoading: Bool
    let onGetResult: (String) -> Void
    let onCancel: () -> Void

    func makeCoordinator() -> Coordinator {
        Coordinator(self)
    }

    func makeUIView(context: Context) -> WKWebView {
        // Configure WKWebView with message handler
        let configuration = WKWebViewConfiguration()
        let contentController = WKUserContentController()

        // Add message handler for cancel action
        contentController.add(context.coordinator, name: "cancelHandler")

        // Inject JavaScript to intercept ONLY cancel button click
        let script = WKUserScript(
            source: """
                (function() {
                    // Store the original loginCancel function if it exists
                    const originalLoginCancel = window.loginCancel;

                    // Override the loginCancel function
                    window.loginCancel = function(event) {
                        event.preventDefault();
                        event.stopPropagation();

                        // Notify Swift about cancel
                        window.webkit.messageHandlers.cancelHandler.postMessage('cancel');

                        // Call original function if it existed
                        if (originalLoginCancel && typeof originalLoginCancel === 'function') {
                            originalLoginCancel.call(this, event);
                        }

                        return false;
                    };

                    // More specific: Only intercept buttons with onclick="loginCancel"
                    document.addEventListener('click', function(e) {
                        const button = e.target.closest('button');

                        // Check if it's specifically the cancel button
                        if (button &&
                            button.classList.contains('button-item-flat') &&
                            button.getAttribute('onclick') &&
                            button.getAttribute('onclick').includes('loginCancel')) {

                            e.preventDefault();
                            e.stopPropagation();
                            window.webkit.messageHandlers.cancelHandler.postMessage('cancel');
                        }
                    }, true);
                })();
                """,
            injectionTime: .atDocumentEnd,
            forMainFrameOnly: true
        )

        contentController.addUserScript(script)
        configuration.userContentController = contentController

        let webView = WKWebView(frame: .zero, configuration: configuration)
        webView.backgroundColor = .white
        webView.isOpaque = true
        webView.navigationDelegate = context.coordinator

        if let url = URL(string: urlString) {
            let request = URLRequest(url: url)
            webView.load(request)
        }

        return webView
    }

    func updateUIView(_ webView: WKWebView, context: Context) {
        if let currentURL = webView.url?.absoluteString,
            currentURL != urlString,
            let url = URL(string: urlString)
        {
            let request = URLRequest(url: url)
            webView.load(request)
        }
    }

    // MARK: - Coordinator
    class Coordinator: NSObject, WKNavigationDelegate, WKScriptMessageHandler {
        let parent: WebView

        init(_ parent: WebView) {
            self.parent = parent
        }

        // Handle messages from JavaScript
        func userContentController(
            _ userContentController: WKUserContentController,
            didReceive message: WKScriptMessage
        ) {
            if message.name == "cancelHandler", message.body as? String == "cancel" {
                DispatchQueue.main.async {
                    self.parent.onCancel()
                }
            }
        }

        func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
            parent.isLoading = true
        }

        func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
            parent.isLoading = false
        }

        func webView(
            _ webView: WKWebView,
            decidePolicyFor navigationAction: WKNavigationAction,
            decisionHandler: @escaping (WKNavigationActionPolicy) -> Void
        ) {
            if let url = navigationAction.request.url {
                // Extract fragment (hash) from URL
                if let fragment = url.fragment, fragment.hasPrefix("tgAuthResult=") {
                    let jsonPart = fragment.replacingOccurrences(of: "tgAuthResult=", with: "")
                    parent.onGetResult(jsonPart)
                }
            }

            decisionHandler(.allow)
        }
    }

    private func clearCookies() {
        let dataStore = WKWebsiteDataStore.default()
        dataStore.fetchDataRecords(ofTypes: WKWebsiteDataStore.allWebsiteDataTypes()) { records in
            dataStore.removeData(
                ofTypes: WKWebsiteDataStore.allWebsiteDataTypes(),
                for: records,
                completionHandler: {}
            )
        }
    }
}

// MARK: - Preview
#Preview {
    TelegramLoginScreen(
        onGetResult: { _ in },
        onCancel: {}
    )
}
